### ISSUE TITLE
DSpace 7: Cannot delete multiple bitstreams at the same time

### ISSUE THREAD

**Issue by pongtawat** [DSpace/DSpace#8029]
**Describe the bug**
In item edit page, only one bitstream can be deleted at a time. If we delete multiple bitstreams at once, only the first on get properly deleted, the rest got marked as deleted in bitstream table but still show up in item page.

**To Reproduce**
Steps to reproduce the behavior: (tested on DSpace 7.1 on demo7 site)
1. Add a few bitstreams to an item
2. Go to item edit page -> Bitstream
3. Select multiple bitstreams for deletion
4. Click save
5. Only the first bitstream got properly deleted.
6. The rest of the bitstreams still show up, but cannot be downloaded.

**Expected behavior**
All selected bitstreams are properly deleted.

**More information**
- The first bitstream got marked as deleted and got deleted from bundle2bitstream table.
- The other bitstreams got marked as deleted but are still there in bundle2bitstream table. 
- We see the following Exception in log file when tested on our installation (7.0):

`
javax.persistence.OptimisticLockException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update public.bundle2bitstream set bitstream_id='a610b77b-92fa-460a-8d0b-9ea8b63ee95a'::uuid where bundle_id='9e30efa6-84db-48a3-97ea-5c6f923aa901'::uuid and bitstream_order=131
        at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:238) ~[hibernate-core-5.4.10.Final.jar:5.4.10.Final]
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:93) ~[hibernate-core-5.4.10.Final.jar:5.4.10.Final]
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181) ~[hibernate-core-5.4.10.Final.jar:5.4.10.Final]
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188) ~[hibernate-core-5.4.10.Final.jar:5.4.10.Final]
        at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1348) ~[hibernate-core-5.4.10.Final.jar:5.4.10.Final]
        at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1331) ~[hibernate-core-5.4.10.Final.jar:5.4.10.Final]
        at jdk.internal.reflect.GeneratedMethodAccessor306.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.hibernate.context.internal.ThreadLocalSessionContext$TransactionProtectionWrapper.invoke(ThreadLocalSessionContext.java:351) ~[hibernate-core-5.4.10.Final.jar:5.4.10.Final]
        at com.sun.proxy.$Proxy185.flush(Unknown Source) ~[?:?]
        at org.dspace.core.HibernateDBConnection.commit(HibernateDBConnection.java:159) ~[dspace-api-7.0.jar:7.0]
        at org.dspace.core.Context.commit(Context.java:432) ~[dspace-api-7.0.jar:7.0]
[...]
`

**Comment by tdonohue** [comment]
I'm able to reproduce this on 7.1.1, although I *do not* see the exception.  Instead, I see this:
* I started with an Item with 4 bitstreams
* I selected 3 of those bitstreams for deletion, and clicked save.
* In my browser's dev tools, I see 3 `DELETE` calls are sent to the backend, and they all return 204.  So, they claim all 3 deletions succeeded.
* However, in the UI, only the first bitstream disappears and the other 2 remain (just as described above).
* If I query the REST API directly for that Item (i.e. `http://localhost:8080/server/api/core/items/[uuid]`), I see that it's "ORIGINAL" bundle still claims to have 3 remaining Bitstreams.
* But, if I query the REST API for the other 2 deleted bitstreams, they are not found. 

So, in summary, it appears all 3 Bitstreams end up deleted. However, the last two are still linked to the item somehow (perhaps via that `bundle2bitstream` table), so they still "appear" as though they exist.

Pulling this over to the 7.3 board in the hopes we can get this solved soon.  Needs a volunteer though.

**Comment by tdonohue** [comment]
@BrunoRoemers : This ticket was moved to 7.4, but I can assign it to you for the release, if you wish to continue work on it.  I honestly haven't had a chance to dig deeply on your ideas, but they do sound like reasonable approaches to investigate.  

I think we'd need to balance the need for multi-deletions (which likely are not frequent themselves) with ensuring good performance (I like the idea of a `@Where` clause but not sure if that'd have an impact on performance) and data integrity.

In any case, good ideas so far & let me know if you want to work more on this for 7.4

**Comment by bkelm** [comment]
So out of curiosity, since we just ran into this on 7.4, is there a specific fix to get rid of the references to the files that no longer exist?
